Added interfaces for Gobblin broker.

ibuenros opened the pull request:
This PR just adds the interfaces for Gobblin broker.

ibuenros commented:
@chavdar @ydai1124 can you review? Thanks!

chavdar commented on gobblin-utility/src/main/java/gobblin/broker/iface/ScopedSharedResourceFactory.java:
```diff
@@ -0,0 +1,25 @@
+  /**
+   * Create an object for the provided {@link SharedResourceKey}, with the provided configuration.
+   */
+  T createResource(ScopedGobblinBroker broker, KeyedScopedConfigView<?, K> config);
```
Why does the factory need an instance of the broker? Can't it just get a scope?

ibuenros replied:
Because the factory may itself need a shared resource. For example, a limiter factory needs to acquire a restli client which is shared by all limiters.

chavdar commented on gobblin-utility/src/main/java/gobblin/broker/iface/SharedResourceFactory.java:
```diff
@@ -0,0 +1,26 @@
+ * @param <T> The type of objects built by this factory.
+ * @param <K> The key used to create objects by this factory.
+ */
+public interface SharedResourceFactory<T, K extends SharedResourceKey> {
```
This interface is strange. It does not define any "factory" methods and does not use any of the generic parameters.

chavdar commented on gobblin-utility/src/main/java/gobblin/broker/iface/UnscopedSharedResourceFactory.java:
```diff
@@ -0,0 +1,25 @@
+ * @param <T> type of objects this {@link SharedResourceFactory} creates.
+ * @param <K> type of {@link SharedResourceKey} this factory uses.
+ */
+public interface UnscopedSharedResourceFactory<T, K extends SharedResourceKey> extends SharedResourceFactory<T, K> {
```
Can you think if we can avoid the parallel class hierarchies for scoped and unscoped?

ydai1124 commented on gobblin-utility/src/main/java/gobblin/broker/iface/KeyedConfigView.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Provides a view of a {@link Config} specific to a {@link SharedResourceKey} and {@link SharedResourceFactory}.
+ */
+public interface KeyedConfigView<K extends SharedResourceKey> extends ConfigView {
```
What is the interface for? What are other use cases besides `KeyedScopedConfigView`?

ydai1124 commented on gobblin-utility/src/main/java/gobblin/broker/iface/GobblinBroker.java:
```diff
@@ -0,0 +1,90 @@
+  /**
+   * @return The lowest {@link Scope} for which this {@link GobblinBroker} can provide a {@link ScopedGobblinBroker}.
+   */
+  Scope<S> leafScope();
```
Is `leafScope()` used for the callers to check if the requested `scopeType` is provided? Then why not simply provide a collection of all the `scopeType`s that are supported?

ibuenros commented:
@chavdar @ydai1124 addressed all comments.
I simplified the interfaces by removing scoped parallels of many things (although scoped and unscoped factories are still separate).

chavdar commented on gobblin-utility/src/main/java/gobblin/broker/iface/ScopeType.java:
```diff
@@ -0,0 +1,55 @@
+ * Represents a DAG of scope types.
+ *
+ * For example, the topology of a distributed application might have scopes as follows:
+ * GLOBAL -> INSTANCE --> JOB       --> TASK
```
nit: It is useful to enclose that in a &lt;pre&gt; tag otherwise, HTML will mess up the alignment.
